docs/internal/capabilities.md

```diff
@@ -0,0 +1,210 @@
+  Once the `iso` region is destroyed, all its `mut` references are invalidated.
+* Read-only (`readonly`).
+  Behaviour-internal references to objects providing local immutability (ie. restrict write access) through this reference.
+  Other mutable references can still modify the object, but only local to the same behaviour, so there are no race conditions.
+* Immutable (`imm`).
+  Global immutability, guarantees that no behaviour will be able to write through references to the same objects.
+  Isolated regions can be entirely frozen into immutable regions.
+  Mutable references are frozen together with the region but can't themselves be frozen directly.
+  Frozen regions cannot be unfrozen and can be freely shared across multiple behaviours.
+
+## Regions
+
+Asynchronous tasks in Verona are performed by behaviours (`when`) on regions.
+The language semantics guarantee (via `iso` pointers) that, at any time, only a single behaviour has access to a region.
+
+Regions are portions of memory dominated by an isolated (`iso`) pointer, with any number of mutable (`mut`) pointers in the same region, accessible only through the `iso` pointer.
+Regions can be implemented in different ways, with different semantics.
+
+For example, they can have different memory management strategies:
+* Garbage collected: Internal objects are created, used and marked for destruction, with a collector freeing resources later.
+* Reference counted: Objects have internal counts for each reference and when the count reaches zero, they're immediately deallocated.
+* Arena model: Objects are allocated and used, but only freed on the destruction of the region itself.
+
+They can also refer to how the code is executed:
+* Native: Verona code calling Verona code, fully checked and safe.
+* Sandbox: Verona code calling an external language module (ex. a C++ library), with process isolation.
+
+Sandboxes themselves can also be of different types, for example:
+* Child process: Forks a new process, shared memory, calls through RPC.
+* WebASM: Isolates OS functionality, memory access, etc.
+* Remote execution: For use with different technologies (ex. GPU) or architectures (ex. connected embedded board).
+
+None of those lists are complete, they're just examples of what can be done.
+
+## Assignment
+
+Capabilities change the semantics of assignments in the language.
+
+**Isolated objects** can only be used through a single reference, so assigning an `iso` region to a new variable moves the reference and invalidates the previous one.
+
+```ts
+var foo : Foo & iso = Foo; // plus undecided syntax for "in new region type X"
+foo.doSomething();
+...
+var bar = foo; // bar now holds the *only* reference to foo
+bar.somethingElse(); // OK: bar is a reference to the same object as foo
+foo.cleanup(); // ERROR: foo is not a valid reference any more
+```
+
+**Mutable objects** can alias each other, so assigning a `mut` variable to another creates a (different) reference to the same variable.
```

Comment:
This is where it gets fuzzy for me, because the stack has a different lifetime than regions.

There are two issues here:
1. Compile time guarantees by the type system: if the object is never accessed after it's invalidated (for whatever reason), the code is "correct" and regardless of how we implement that at run-time, it's fine.
2. Run-time semantics: if the compiler said no one access the object after it's been made invalid, then stack lifetime is irrelevant, ie. all stack frames can have the "arena" model.

What cannot happen is functions trying to access objects in other frames that they don't own. Previous frames could be fine because their lifetime is strictly longer than the current's, but it'd be up for the compiler to prove that the access is correct via its own region's lifetime (which I'm not sure it's always possible).